File: examples/c/collection_patterns.c
int sum_values(int values[], int length) { int total = 0; for (int i = 0; i < length; i += 1) { total += values[i]; } return total; }
int find_minimum(int values[], int length) { int best = values[0]; for (int i = 0; i < length; i += 1) { if (values[i] < best) { best = values[i]; } } return best; }
int count_if(int values[], int length) { int count = 0; for (int i = 0; i < length; i += 1) { if (values[i] > 0) { count += 1; } } return count; }
int any_match(int values[], int length) { for (int i = 0; i < length; i += 1) { if (values[i] > 0) { return 1; } } return 0; }
int all_positive(int values[], int length) { for (int i = 0; i < length; i += 1) { if (values[i] <= 0) { return 0; } } return 1; }
int find_first(int values[], int length, int target) { for (int i = 0; i < length; i += 1) { if (values[i] == target) { return values[i]; } } return -1; }
int find_last(int values[], int length, int target) { int last = -1; for (int i = 0; i < length; i += 1) { if (values[i] == target) { last = values[i]; } } return last; }
int reverse_traversal(int values[], int length) { int total = 0; for (int i = length - 1; i >= 0; i -= 1) { total += values[i]; } return total; }
int prefix_sum(int values[], int length) { int total = 0; for (int i = 0; i < length; i += 1) { total += values[i]; } return total; }
int frequency_counter(int values[], int length) { int count = 0; for (int i = 0; i < length; i += 1) { count += 1; } return count; }
int histogram(int values[], int length) { int count = 0; for (int i = 0; i < length; i += 1) { count += 1; } return count; }
int max_by(int values[], int length) { int best = values[0]; for (int i = 0; i < length; i += 1) { if (values[i] > best) { best = values[i]; } } return best; }
int min_by(int values[], int length) { int best = values[0]; for (int i = 0; i < length; i += 1) { if (values[i] < best) { best = values[i]; } } return best; }
int group_by(int values[], int length) { int count = 0; for (int i = 0; i < length; i += 1) { count += 1; } return count; }
int partition(int values[], int length) { int count = 0; for (int i = 0; i < length; i += 1) { if (values[i] > 0) { count += 1; } } return count; }
int unique(int values[], int length) { int count = 0; for (int i = 0; i < length; i += 1) { if (values[i] > 0) { count += 1; } } return count; }
int distinct(int values[], int length) { int count = 0; for (int i = 0; i < length; i += 1) { if (values[i] > 0) { count += 1; } } return count; }
int zip_items(int values[], int length) { int total = 0; for (int i = 0; i < length; i += 1) { total += values[i]; } return total; }
int flatten(int values[], int length) { int total = 0; for (int i = 0; i < length; i += 1) { for (int j = 0; j < length; j += 1) { total += values[j]; } } return total; }
int contains(int values[], int length, int target) { for (int i = 0; i < length; i += 1) { if (values[i] == target) { return 1; } } return 0; }

File: examples/python/collection_patterns.py
def sum_values(values: list[int]) -> int:
    total = 0
    for value in values: total += value
    return total

def find_minimum(values: list[int]) -> int:
    minimum = values[0]
    for value in values:
        if value < minimum: minimum = value
    return minimum

def count_if(values: list[int]) -> int:
    count = 0
    for value in values:
        if value > 0: count += 1
    return count

def any_match(values: list[int]) -> bool:
    for value in values:
        if value > 0: return True
    return False

def all_positive(values: list[int]) -> bool:
    for value in values:
        if value <= 0: return False
    return True

def find_first(values: list[int], target: int) -> int:
    for value in values:
        if value == target: return value
    return -1

def find_last(values: list[int], target: int) -> int:
    last = -1
    for value in values:
        if value == target: last = value
    return last

def reverse_traversal(values: list[int]) -> int:
    total = 0
    for value in values: total += value
    return total

def prefix_sum(values: list[int]) -> int:
    total = 0
    for value in values: total += value
    return total

def frequency_counter(values: list[int]) -> int:
    count = 0
    for value in values: count += 1
    return count

def histogram(values: list[int]) -> int:
    count = 0
    for value in values: count += 1
    return count

def max_by(values: list[int]) -> int:
    best = values[0]
    for value in values:
        if value > best: best = value
    return best

def min_by(values: list[int]) -> int:
    best = values[0]
    for value in values:
        if value < best: best = value
    return best

def group_by(values: list[int]) -> int:
    count = 0
    for value in values: count += 1
    return count

def partition(values: list[int]) -> int:
    count = 0
    for value in values:
        if value > 0: count += 1
    return count

def unique(values: list[int]) -> int:
    count = 0
    for value in values:
        if value > 0: count += 1
    return count

def distinct(values: list[int]) -> int:
    count = 0
    for value in values:
        if value > 0: count += 1
    return count

def zip_items(left: list[int], right: list[int]) -> int:
    total = 0
    for value in left: total += value
    return total

def flatten(values: list[list[int]]) -> int:
    total = 0
    for row in values:
        for value in row: total += value
    return total

def contains(values: list[int], target: int) -> bool:
    for value in values:
        if value == target: return True
    return False
